lib/services/reconciler.go

```diff
@@ -125,62 +135,82 @@ func (r *Reconciler) Reconcile(ctx context.Context) error {
+	name := r.cfg.GetName(newT)
 	// First see if the resource is already registered and if not, whether it
 	// matches the selector labels and should be registered.
-	registered := currentResources[new.GetName()]
-	if registered == nil {
-		if r.cfg.Matcher(new) {
-			r.log.Infof("%v %v matches, creating.", new.GetKind(), new.GetName())
-			if err := r.cfg.OnCreate(ctx, new); err != nil {
-				return trace.Wrap(err, "failed to create %v %v", new.GetKind(), new.GetName())
+	registered, ok := currentResources[name]
+	if !ok {
+		kind := r.cfg.GetKind(registered)
+		if r.cfg.Matcher(newT) {
+			r.log.Infof("%v %v matches, creating.", kind, name)
+			if err := r.cfg.OnCreate(ctx, newT); err != nil {
+				return trace.Wrap(err, "failed to create %v %v", kind, name)
 			}
 			return nil
 		}
-		r.log.Debugf("%v %v doesn't match, not creating.", new.GetKind(), new.GetName())
+		r.log.Debugf("%v %v doesn't match, not creating.", kind, name)
 		return nil
 	}
 
 	// Don't overwrite resource of a different origin (e.g., keep static resource from config and ignore dynamic resource)
-	if registered.Origin() != new.Origin() {
-		r.log.Warnf("%v has different origin (%v vs %v), not updating.", new.GetName(),
-			new.Origin(), registered.Origin())
+	registeredOrigin := r.cfg.GetOrigin(registered)
+	newOrigin := r.cfg.GetOrigin(newT)
+	if registeredOrigin != newOrigin {
+		r.log.Warnf("%v has different origin (%v vs %v), not updating.", name, newOrigin, registeredOrigin)
 		return nil
 	}
 
 	// If the resource is already registered but was updated, see if its
 	// labels still match.
-	if CompareResources(new, registered) != Equal {
-		if r.cfg.Matcher(new) {
-			r.log.Infof("%v %v updated, updating.", new.GetKind(), new.GetName())
-			if err := r.cfg.OnUpdate(ctx, new); err != nil {
-				return trace.Wrap(err, "failed to update %v %v", new.GetKind(), new.GetName())
+	kind := r.cfg.GetKind(registered)
+	if CompareResources(newT, registered) != Equal {
+		if r.cfg.Matcher(newT) {
+			r.log.Infof("%v %v updated, updating.", kind, name)
+			if err := r.cfg.OnUpdate(ctx, newT); err != nil {
+				return trace.Wrap(err, "failed to update %v %v", kind, name)
 			}
 			return nil
 		}
-		r.log.Infof("%v %v updated and no longer matches, deleting.", new.GetKind(), new.GetName())
+		r.log.Infof("%v %v updated and no longer matches, deleting.", kind, name)
 		if err := r.cfg.OnDelete(ctx, registered); err != nil {
-			return trace.Wrap(err, "failed to delete %v %v", new.GetKind(), new.GetName())
+			return trace.Wrap(err, "failed to delete %v %v", kind, name)
 		}
 		return nil
 	}
 
-	r.log.Debugf("%v %v is already registered.", new.GetKind(), new.GetName())
+	r.log.Debugf("%v %v is already registered.", kind, name)
 	return nil
 }
+
+// ToMap returns these databases as a map keyed by database name.
+func ToMap[T any](r []T, key func(T) string) map[string]T {
+	rm := make(map[string]T, len(r))
```

Comment:
nit:
```suggestion
	m := make(map[string]T, len(r))
```

`rm` reads like the command `rm`, which gave me pause for a second in the lines below.